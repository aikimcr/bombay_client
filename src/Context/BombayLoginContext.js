import { createContext } from "react";

const BombayLoginContext = createContext({
  loggedIn: false,
  showLoginForm: false,
  setLoggedIn: null,
  setShowLogin: null,
});

export default BombayLoginContext;
