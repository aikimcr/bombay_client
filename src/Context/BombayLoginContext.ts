import { createContext, Dispatch } from "react";

interface BombayLoginContextValues {
  loggedIn: boolean;
  setLoggedIn?: Dispatch<boolean>;
}

export const BombayLoginContext = createContext({
  loggedIn: false,
  setLoggedIn: null,
});

export default BombayLoginContext;
