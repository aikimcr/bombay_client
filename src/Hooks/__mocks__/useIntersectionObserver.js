import { useEffect, useRef } from "react";

export class mockEntry {
  constructor(target, isIntersecting = false) {
    this.target = target;
    this.isIntersecting = isIntersecting;
    this.isVisible = false;
  }
}

export class mockObserver {
  static observers = [];

  constructor(callback, options) {
    this.callback = callback;
    this.options = options;
    this.entries = [];
    this.constructor.observers.push(this);
  }

  observe(target) {}

  unobserve(target) {}

  disconnect() {}

  _fireIntersect(target, isIntersecting = true) {
    const entry = new mockEntry(target, isIntersecting);
    this.callback([entry], this);
  }
}

export function useIntersectionObserver(topRef, callback, options = {}) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) return;

    if (observer.current == null) {
      observer.current = new mockObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [observer.current, topRef]);

  return observer;
}

export default useIntersectionObserver;
