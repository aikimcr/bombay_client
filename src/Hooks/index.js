export * from './useLoginTracking';
export * from './useModelCollection';
