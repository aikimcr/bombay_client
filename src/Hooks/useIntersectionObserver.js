import { useEffect, useRef } from "react";

export function useIntersectionObserver(topRef, callback, options = {}) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) return;

    if (observer.current == null) {
      observer.current = new IntersectionObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [callback, options, topRef]);

  return observer;
}

export default useIntersectionObserver;
