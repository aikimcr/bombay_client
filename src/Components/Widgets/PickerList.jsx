import { createRef, useContext } from "react";
import PropTypes from "prop-types";

import BombayLoginContext from "../../Context/BombayLoginContext";
import useModelCollection from "../../Hooks/useModelCollection";

function PickerList({ pickModel, isOpen, collectionClass }) {
  const topRef = createRef();

  const loginState = useContext(BombayLoginContext);

  const [listCollection] = useModelCollection({
    CollectionClass: collectionClass,
    topRef,
    loginState,
  });

  function clickHandler(evt, model) {
    evt.preventDefault();
    pickModel(model);
  }

  if (!isOpen) return null;

  return (
    <div className="picker-component">
      <div className="list-container" ref={topRef}>
        <ul className="picker-list">
          {listCollection?.current == null
            ? ""
            : listCollection.current.map((model) => {
                const key = `mode-list-${model.get("id")}`;
                return (
                  <li
                    className="picker-item"
                    key={key}
                    data-model-id={model.get("id")}
                    onClick={(evt) => {
                      clickHandler(evt, model);
                    }}
                  >
                    {model.get("name")}
                  </li>
                );
              })}
        </ul>
      </div>
    </div>
  );
}

PickerList.propTypes = {
  pickModel: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  collectionClass: PropTypes.func.isRequired,
};

export default PickerList;
