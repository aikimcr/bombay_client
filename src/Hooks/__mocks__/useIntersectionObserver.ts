import { useEffect, useRef, RefObject } from 'react';

import { IntersectionObserverOptions } from '../useIntersectionObserver';
export class mockEntry {
  boundingClientRect: DOMRectReadOnly;
  intersectionRatio: number;
  intersectionRect: DOMRectReadOnly;
  isIntersecting: boolean;
  rootBounds: DOMRectReadOnly;
  target: Element;
  time: number;
  isVisible: boolean;

  constructor(target: Element, isIntersecting = false) {
    this.target = target;
    this.isIntersecting = isIntersecting;
    this.isVisible = false;
  }
}

export class mockObserver {
  root: Element;
  rootMargin: string;
  thresholds: readonly number[];
  callback: IntersectionObserverCallback;
  options: IntersectionObserverOptions;
  entries: mockEntry[];

  static observers: mockObserver[] = [];

  constructor(
    callback: IntersectionObserverCallback,
    options: IntersectionObserverOptions,
  ) {
    this.callback = callback;
    this.options = options;
    this.entries = [];
    mockObserver.observers.push(this);
  }

  observe(_target: Element) {}

  unobserve(_target: Element) {}

  takeRecords() {
    return this.entries;
  }

  disconnect() {}

  _fireIntersect(target: Element, isIntersecting = true) {
    const entry = new mockEntry(target, isIntersecting);
    this.callback([entry], this);
  }
}

export function useIntersectionObserver(
  topRef: RefObject<Element>,
  callback: IntersectionObserverCallback,
  options: IntersectionObserverOptions = {},
) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) {
      return;
    }

    if (observer.current == null) {
      observer.current = new mockObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [observer.current, topRef]);

  return observer;
}

export default useIntersectionObserver;
