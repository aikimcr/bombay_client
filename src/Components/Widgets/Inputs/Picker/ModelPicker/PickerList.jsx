import React, { createRef, useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import BombayLoginContext from '../../../../../Context/BombayLoginContext';
import { useModelCollection } from '../../../../../Hooks/useModelCollection';

export function PickerList({ pickModel, isOpen, initialCollection }) {
  const topRef = createRef();

  const loginState = useContext(BombayLoginContext);

  const [isMounted, setIsMounted] = useState(false);
  const [listCollection] = useState(initialCollection);
  const { refreshCollection } = useModelCollection({
    initialCollection: listCollection,
    topRef,
    isMounted,
  });

  useEffect(() => {
    setIsMounted(true);
  }, []);

  function clickHandler(evt, model) {
    evt.preventDefault();
    pickModel(model);
  }

  if (!isOpen) return null;

  return (
    <div className="picker-component" data-testid="picker-component">
      <div className="list-container" ref={topRef}>
        <ul className="picker-list">
          {!listCollection
            ? ''
            : listCollection.map((model) => {
                const key = `mode-list-${model.id}`;
                return (
                  <li
                    className="picker-item"
                    data-testid="picker-item"
                    key={key}
                    data-model-id={model.id}
                    onClick={(evt) => {
                      clickHandler(evt, model);
                    }}
                  >
                    {model.name}
                  </li>
                );
              })}
        </ul>
      </div>
    </div>
  );
}

PickerList.propTypes = {
  pickModel: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  collectionClass: PropTypes.func.isRequired,
};

export default PickerList;
