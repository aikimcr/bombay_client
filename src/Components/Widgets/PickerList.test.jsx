import { useState } from "react";

import { act, render, screen } from "@testing-library/react";

import * as NetworkLogin from "../../Network/Login";
jest.mock("../Network/Login");

import * as Network from "../../Network/Network";
import * as mockObserver from "../../Hooks/useIntersectionObserver";

import { makeModels } from "../../testHelpers/modelTools";

import BombayLoginContext from "../../Context/BombayLoginContext";

import ModelBase from "../../model/ModelBase";
import CollectionBase from "../../model/CollectionBase";

class TestModel extends ModelBase {}

class TestCollection extends CollectionBase {
  constructor(options = {}) {
    super("/table1", { ...options, modelClass: TestModel });
  }
}

import PickerList from "./PickerList.jsx";

function TestWrapper({ loggedIn, showLoginForm, children }) {
  const [loginState, setLoginState] = useState({ loggedIn, showLoginForm });

  return (
    <BombayLoginContext.Provider value={loginState}>
      {children}
    </BombayLoginContext.Provider>
  );
}

jest.useFakeTimers();

const testToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJGREM4MTEzOCIsInVzZXIiOnsiaWQiOjEsIm5hbWUiOiJhZG1pbiIsImFkbWluIjpmYWxzZX0sImlhdCI6MTY2NTk2NTA5OX0.2vz14X7Tm-oFlyOa7dcAF-5y5ympi_UlWyJNxO4xyS4";

function setupLogin(loggedIn = true, token = testToken) {
  const loginPromise = NetworkLogin._setupMocks();
  loginPromise.resolve({ loggedIn, token });
}

beforeEach(() => {
  localStorage.setItem("jwttoken", testToken);
});

afterEach(() => {
  while (mockObserver.mockObserver.observers.length > 0) {
    mockObserver.mockObserver.observers.pop();
  }

  localStorage.removeItem("jwttoken");
});

it("should show the list", async () => {
  setupLogin();
  const pickModel = jest.fn();

  const { resolve } = Network._setupMocks();

  const { asFragment } = render(
    <TestWrapper loggedIn={true} showLoginForm={false}>
      <PickerList
        pickModel={pickModel}
        isOpen={true}
        collectionClass={TestCollection}
      />
    </TestWrapper>,
  );

  const [fetchBody, models] = makeModels(10, {});

  await act(async () => {
    resolve(fetchBody);
  });

  expect(asFragment).toMatchSnapshot();

  expect(mockObserver.mockObserver.observers.length).toBe(1);
  expect(Network.getFromURLString).toBeCalledTimes(1);
  expect(Network.getFromURLString).toBeCalledWith(
    "http://localhost:2001/xyzzy/table1",
  );
  expect(Network.getFromURLString.mock.results[0].value).resolves.toEqual(
    fetchBody,
  );

  const listElements = asFragment().querySelectorAll("li.picker-item");
  expect(listElements).toHaveLength(10);

  const el = screen.queryByText(models[1].get("name"));
  el.click();

  expect(pickModel).toBeCalled();
  expect(pickModel).toBeCalledWith(models[1]);
});

it("should show an empty list", async () => {
  setupLogin();
  const pickModel = jest.fn();

  const { reject } = Network._setupMocks();

  const { container } = render(
    <TestWrapper loggedIn={true} showLoginForm={false}>
      <PickerList
        pickModel={pickModel}
        isOpen={true}
        collectionClass={TestCollection}
      />
    </TestWrapper>,
  );

  await act(async () => {
    reject({ status: 404, message: "Not Found" });
  });

  const listElements = container.querySelectorAll("li.picker-item");
  expect(listElements).toHaveLength(0);
});
