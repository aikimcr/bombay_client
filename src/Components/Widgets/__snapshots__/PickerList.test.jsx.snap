// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="picker-component"
    data-testid="picker-component"
  >
    <div
      class="list-container"
    >
      <ul
        class="picker-list"
      >
        <li
          class="picker-item"
          data-model-id="1"
          data-testid="picker-item"
        >
          fellow
        </li>
        <li
          class="picker-item"
          data-model-id="2"
          data-testid="picker-item"
        >
          property
        </li>
        <li
          class="picker-item"
          data-model-id="3"
          data-testid="picker-item"
        >
          yet
        </li>
        <li
          class="picker-item"
          data-model-id="4"
          data-testid="picker-item"
        >
          article
        </li>
        <li
          class="picker-item"
          data-model-id="5"
          data-testid="picker-item"
        >
          held
        </li>
        <li
          class="picker-item"
          data-model-id="6"
          data-testid="picker-item"
        >
          transportation
        </li>
        <li
          class="picker-item"
          data-model-id="7"
          data-testid="picker-item"
        >
          scene
        </li>
        <li
          class="picker-item"
          data-model-id="8"
          data-testid="picker-item"
        >
          current
        </li>
        <li
          class="picker-item"
          data-model-id="9"
          data-testid="picker-item"
        >
          life
        </li>
        <li
          class="picker-item"
          data-model-id="10"
          data-testid="picker-item"
        >
          printed
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
